src/sentry/api/endpoints/internal/rpc.py

```diff
@@ -0,0 +1,74 @@
+from __future__ import annotations
+
+from django.conf import settings
+from OpenSSL.crypto import X509Name
+from rest_framework import permissions
+from rest_framework.exceptions import NotFound
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from sentry.api.authentication import MutualTlsAuthentication
+from sentry.api.base import Endpoint, control_silo_endpoint, region_silo_endpoint
+from sentry.services.hybrid_cloud.rpc import ServiceEndpoint, endpoint_registry
+from sentry.services.hybrid_cloud.rpc.endpoints import MethodEndpoint
+from sentry.silo import SiloLimit, SiloMode
+
+
+class SiloPermission(permissions.BasePermission):
+    def has_permission(self, request: Request, view: RpcEndpoint):
+        auth: X509Name = request.auth
+        if not isinstance(request.auth, X509Name):
+            common_name: str = auth.commonName
+            is_control_silo = common_name == settings.CONTROL_SILO_COMMON_NAME
+
+            # Region silo RPCs cannot communicate with each other; they may only be
+            # contacted on behalf of the control silo.
+            if view.silo_mode() == SiloMode.REGION:
+                return is_control_silo
+
+            return True
+        return False
+
+
+class RpcEndpoint(Endpoint):
+    authentication_classes = (MutualTlsAuthentication,)
+    permission_classes = (SiloPermission,)
+
+    def silo_mode(self) -> SiloMode:
```

Comment:
I don't quite understand the motivation here. `silo_limit.modes` clearly can hold more than one value, but this method returns the first one returned. Yet there is a big difference in the `SiloPermission` between `CONTROL` and `REGION` and `MONOLITH` appears to be an ignored value. If this logic is correct, then it needs clear documentation. I cannot quite tell how this is supposed to behave in the presence of multiple values.

Reply:
The goal is that RPC endpoints explicitly only support one silo mode, period.  That's kind of the goal.  We do have some existing view endpoints (catchall) we might want to support both, but the intention here was for RPC endpoints to always be explicit about their alignment.  Motivating reason was so that configuration issues and authentication are always clear -- no intention for regions to speak between each other.

Reply:
`silo_limit.modes` implementation is honestly likely to change.  It will hold a single value in practice.  This is really an artifact of existing implementation but can be changed.

Reply:
Right. Ideally we can get this to be a singular value then and avoid an invalid variant in the enum.